internal/debug/flags.go

```diff
@@ -148,44 +183,66 @@ func init() {
 // Setup initializes profiling and logging based on the CLI flags.
 // It should be called as early as possible in the program.
 func Setup(ctx *cli.Context) error {
-	logFile := ctx.String(logFileFlag.Name)
-	useColor := logFile == "" && os.Getenv("TERM") != "dumb" && (isatty.IsTerminal(os.Stderr.Fd()) || isatty.IsCygwinTerminal(os.Stderr.Fd()))
-
-	var logfmt log.Format
-	switch ctx.String(logFormatFlag.Name) {
-	case "json":
+	var (
+		logfmt     log.Format
+		output     = io.Writer(os.Stderr)
+		logFmtFlag = ctx.String(logFormatFlag.Name)
+	)
+	switch {
+	case ctx.Bool(logjsonFlag.Name):
+		// Retain backwards compatibility with `--log.json` flag if `--log.format` not set
+		defer log.Warn("The flag '--log.json' is deprecated, please use '--log.format=json' instead")
 		logfmt = log.JSONFormat()
-	case "logfmt":
+	case logFmtFlag == "json":
+		logfmt = log.JSONFormat()
+	case logFmtFlag == "logfmt":
 		logfmt = log.LogfmtFormat()
-	case "terminal":
-		logfmt = log.TerminalFormat(useColor)
-	case "":
-		// Retain backwards compatibility with `--log.json` flag if `--log.format` not set
-		if ctx.Bool(logjsonFlag.Name) {
-			defer log.Warn("The flag '--log.json' is deprecated, please use '--log.format=json' instead")
-			logfmt = log.JSONFormat()
-		} else {
-			logfmt = log.TerminalFormat(useColor)
+	case logFmtFlag == "", logFmtFlag == "terminal":
+		useColor := (isatty.IsTerminal(os.Stderr.Fd()) || isatty.IsCygwinTerminal(os.Stderr.Fd())) && os.Getenv("TERM") != "dumb"
+		if useColor {
+			output = colorable.NewColorableStderr()
 		}
+		logfmt = log.TerminalFormat(useColor)
 	default:
 		// Unknown log format specified
 		return fmt.Errorf("unknown log format: %v", ctx.String(logFormatFlag.Name))
 	}
-
-	if logFile != "" {
-		var err error
-		logOutputStream, err = log.FileHandler(logFile, logfmt)
-		if err != nil {
-			return err
+	var (
+		stdHandler = log.StreamHandler(output, logfmt)
+		ostream    = stdHandler
+		logFile    = ctx.String(logFileFlag.Name)
+		rotation   = ctx.Bool(logRotateFlag.Name)
+	)
+	if len(logFile) > 0 {
+		if err := validateLogLocation(filepath.Dir(logFile)); err != nil {
+			return fmt.Errorf("failed to initiatilize file logger: %v", err)
 		}
-	} else {
-		output := io.Writer(os.Stderr)
-		if useColor {
-			output = colorable.NewColorableStderr()
+	}
+	if rotation {
+		// Lumberjack uses <processname>-lumberjack.log in is.TempDir() if empty.
+		// so typically /tmp/geth-lumberjack.log on linux
+		if len(logFile) > 0 {
+			log.Info("Rotating file logging configured", "location", logFile)
```

Comment:
Perhaps we should defer these? In case `format=json` is specified, it's a bit annoying to have a non-json regular log being spat out before the json-logs. 
```suggestion
			defer log.Info("Rotating file logging configured", "location", logFile)
```

@fjl wdyt?

Reply:
I agree with your reasoning that it's not good to print a message while logging is not fully configured yet. 

I'm also wondering, are these messages really necessary? I think these should be at debug level or something. Geth already prints a ton of INFO messages at startup.

Reply:
> I'm also wondering, are these messages really necessary?

Maybe not... 

Though I generally find it annoying when debugging something that does _not_ work, and it is not possible to see "what does this application _think_ I told it to do". And with rotation, you won't notice that it does (or does not) work until much later, when it fails to rotate. 

Like, if you configure everything about the rotation, but omitted `log.rotate`, then it's better UX if users are able to notice immediately. (and now that I say that, I also feel that we should add "rotate","off" on the other message)

Reply:
Ok, now I do print out the info in case non-default options are used. I'm happy with it now